### PR TITLE
Add markdown editor on conversations and comments

### PR DESCRIPTION
**Pull request by ghost**
On comments :

markdown editor on comments

On a new conversation :

markdown editor on new conversation

On a conversation reply :

markdown editor on conversation reply

Just needs some tests.

### CODE REVIEW

**Comment by AugierLe42e** [comment]
Tests OK. Ready for review.

**Comment by Flaburgan** [comment]
I tested this locally on both Firefox and Chromium and it works well. Nothing special to report. Thank you @AugierLe42e !

**Comment by svbergerem** [comment]
I'd like the comment textarea to behave just like the publisher (well, except for #7245): 
- in write mode with an empty textarea go back to the initial state when a click event is fired outside of the editor (including the height of the textarea)
- keep the editor opened if it is in preview mode or the textarea is not empty

Both is not happening right now: http://videobin.org/+f0n/iv9.ogg

**Review by svbergerem — CHANGES_REQUESTED** [review]
See https://github.com/diaspora/diaspora/pull/7214#issuecomment-266020391.

**Comment by AugierLe42e** [comment]
I made the requested changes. Please let me know if you need anything else.

**Comment by svbergerem** [comment]
> in write mode with an empty textarea go back to the initial state when a click event is fired outside of the editor (including the height of the textarea)

Enter some text in the textarea, delete it and then click outside of the editor. It doesn't go back to the initial state.

**Comment by AugierLe42e** [comment]
Really? Crap, I didn't test this case :/

**Comment by svbergerem** [comment]
Yes, really. This was one of the cases I tested and which was also part of the video I uploaded.

**Comment by AugierLe42e** [comment]
Meh. I didn't do my homework very well this time :(

**Comment by AugierLe42e** [comment]
> Enter some text in the textarea, delete it and then click outside of the editor. It doesn't go back to the initial state.

It works for me unless I've missed something.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 1]
```diff
@@ -9,10 +9,12 @@ app.views.CommentStream = app.views.Base.extend({
   events: {
     "keydown .comment_box": "keyDownOnCommentBox",
     "submit form": "createComment",
-    "focus .comment_box": "commentTextareaFocused",
-    "click .toggle_post_comments": "expandComments"
+    "click .toggle_post_comments": "expandComments",
+    "click form": "openForm"
   },
 
+  mdEditor: undefined,
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: Why did you add this line?

File: app/assets/javascripts/helpers/markdown_editor.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ Diaspora.MarkdownEditor = function(element, opts) {
 
 Diaspora.MarkdownEditor.prototype = {
   constructor: Diaspora.MarkdownEditor,
+  instance: undefined,
```
Comment (svbergerem) [inline app/assets/javascripts/helpers/markdown_editor.js]: Why did you add this line?

File: app/assets/javascripts/helpers/markdown_editor.js [hunk 3]
```diff
@@ -130,6 +131,13 @@ Diaspora.MarkdownEditor.prototype = {
     }
   },
 
+  isPreviewOrTexareaNotEmpty: function() {
+    if (this.instance === undefined) {
+      return false;
+    }
+    return (this.instance.$editor.find(".md-preview").length > 0) || (this.instance.getContent().length > 0);
+  },
+
```
Comment (svbergerem) [inline app/assets/javascripts/helpers/markdown_editor.js]: I can haz tests?

File: app/assets/javascripts/app/views/conversations_inbox_view.js [hunk 4]
```diff
@@ -23,6 +23,7 @@ app.views.ConversationsInbox = Backbone.View.extend({
         self.$el.find("#conversation-show").removeClass("hidden").html(data);
         self.selectConversation(conversationId);
         self.setupConversation();
+        self.conversationForm.renderMarkdownEditor("#conversation-show .conversation-message-text");
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_inbox_view.js]: Please add a test that we call this method.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 5]
```diff
@@ -21,12 +23,20 @@ app.views.CommentStream = app.views.Base.extend({
   setupBindings: function() {
     this.model.comments.bind("add", this.appendComment, this);
     this.model.comments.bind("remove", this.removeComment, this);
+    $(document.body).click(this.onFormBlur.bind(this));
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: Please add a test.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 6]
```diff
@@ -21,12 +23,20 @@ app.views.CommentStream = app.views.Base.extend({
...
     this.model.comments.each(this.appendComment, this);
     this.commentBox = this.$(".comment_box");
     this.commentSubmitButton = this.$("input[name='commit']");
+
+    this.mdEditor = new Diaspora.MarkdownEditor(this.$(".comment_box"), {
+      onPreview: function($mdInstance) {
+        return "<div class='preview-content'>" + app.helpers.textFormatter($mdInstance.getContent()) + "</div>";
+      },
+      onFocus: this.openForm.bind(this)
+    });
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: Please add a test that we set `mdEditor` and create a new `Diaspora.MarkdownEditor`.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 7]
```diff
@@ -134,6 +143,29 @@ app.views.CommentStream = app.views.Base.extend({
...
+  onFormBlur: function(evt) {
+    if (this.mdEditor !== undefined && this.mdEditor.isPreviewOrTexareaNotEmpty()) {
+      return;
+    }
+
+    var $target = $(evt.target);
+    var isForm = $target.hasClass("new-comment") || $target.parents(".new-comment").length !== 0;
+    if (!isForm && !$target.hasClass("focus_comment_textarea")) {
+      this.closeForm();
+    }
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: Tests for these three new methods would be great.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 8]
```diff
@@ -52,11 +62,14 @@ app.views.CommentStream = app.views.Base.extend({
       success: function() {
         this.commentBox.val("");
         this.enableCommentBox();
+        this.mdEditor.hidePreview();
+        this.closeForm();
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: There are already some tests for the behavior of this function, so please add tests for the added behavior.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 9]
```diff
@@ -52,11 +62,14 @@ app.views.CommentStream = app.views.Base.extend({
...
         autosize.update(this.commentBox);
       }.bind(this),
       error: function() {
         this.enableCommentBox();
-        this.commentBox.focus();
+        this.mdEditor.hidePreview();
+        this.openForm();
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/comment_stream_view.js]: There are already some tests for the behavior of this function, so please add tests for the added behavior.

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 10]
```diff
@@ -24,17 +24,33 @@ app.views.ConversationsForm = Backbone.View.extend({
       remoteRoute: {url: "/contacts", extraParameters: "mutual=true"}
     });
 
+    this.newConversationMdEditor = this.renderMarkdownEditor("#new-message-text");
+
+    // Creates another markdown editor in case of displaying another conversation
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: > [...] in case of displaying ~~another~~ conversation

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 11]
```diff
@@ -24,17 +24,33 @@ app.views.ConversationsForm = Backbone.View.extend({
       remoteRoute: {url: "/contacts", extraParameters: "mutual=true"}
     });
 
+    this.newConversationMdEditor = this.renderMarkdownEditor("#new-message-text");
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Test? ;)

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 12]
```diff
@@ -83,6 +99,7 @@ app.views.ConversationsForm = Backbone.View.extend({
   },
 
   conversationCreateSuccess: function(evt, data) {
+    this.newConversationMdEditor.hidePreview();
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: This could need a test.

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 13]
```diff
@@ -24,17 +24,33 @@ app.views.ConversationsForm = Backbone.View.extend({
...
     this.$("form#new-conversation").on("ajax:error", this.conversationCreateError);
   },
 
+  renderMarkdownEditor: function(element) {
+    return new Diaspora.MarkdownEditor($(element), {
+      onPreview: function($mdInstance) {
+        return "<div class='preview-content'>" + app.helpers.textFormatter($mdInstance.getContent()) + "</div>";
+      }
+    });
+  },
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Please add a test.

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 14]
```diff
@@ -24,17 +24,33 @@ app.views.ConversationsForm = Backbone.View.extend({
       remoteRoute: {url: "/contacts", extraParameters: "mutual=true"}
     });
 
+    this.newConversationMdEditor = this.renderMarkdownEditor("#new-message-text");
+
+    // Creates another markdown editor in case of displaying another conversation
+    var responseTextarea = $("#conversation-show .conversation-message-text");
+    if (responseTextarea.length === 1) {
+      this.renderMarkdownEditor(responseTextarea);
+    }
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Guess what: tests would be great! ;)

**Review by svbergerem — CHANGES_REQUESTED** [review]
I CAN HAZ MOAR TESTZ?!

File: app/assets/javascripts/helpers/markdown_editor.js (hunk 3 above)
Comment (AugierLe42e) [inline app/assets/javascripts/helpers/markdown_editor.js]: :3

**Comment by AugierLe42e** [comment]
Let me guess... You want some tests? :p
I'm on it ;)

File: app/assets/javascripts/app/views/comment_stream_view.js (hunk 1 above)
Comment (AugierLe42e) [inline app/assets/javascripts/app/views/comment_stream_view.js]: To be explicit that `mdEditor` should be `undefined` when the object is initialized. I still don't really understand big parts of JS so that may be useless. I don't know.

File: app/assets/javascripts/app/views/conversations_form_view.js (hunk 13 above)
Comment (AugierLe42e) [inline app/assets/javascripts/app/views/conversations_form_view.js]: What exactly do you want to test?

**Comment by AugierLe42e** [comment]
I've written the tests. Ready for another review.

**Comment by svbergerem** [comment]
Sorry, just found two more bugs. (I didn't review the tests, yet)

When you submit a comment with <kbd>ctrl</kbd>+<kbd>enter</kbd> then you will see a new comment textarea to write a new comment. But you won't see the markdown editor unless you click on the comment textarea. So when submitting the comment form and focusing the new comment textarea we should show the markdown editor.

One could try to bind the `focus` event in addition to the `click` event or maybe even instead of it but I haven't tried it myself.

The second bug is a CSS issue. The editor looks great with the original color theme but has some issues in the dark color theme. We have those issues in the comment and conversations form.

Write mode:
write mode

Preview mode:
preview mode

In the normal publisher it looks good:
publisher write mode
publisher preview mode

So the issues I found were:
* Text color of the active tab (“Write” and “Preview”)
* Background color of the textarea
* Background color of the preview

I'd be happy if you could make the comment and conversations forms look like the publisher. (except for the publisher-specific stuff ;))

**Comment by AugierLe42e** [comment]
Thank you for your careful review, @svbergerem. I take care of that.

**Comment by AugierLe42e** [comment]
Ok, CSS is fixed. 

> So when submitting the comment form and focusing the new comment textarea we should show the markdown editor.

That's because the comment area is closed when the submission has succeeded. So, instead, I put a blur in the `closeForm` method.

**Comment by AugierLe42e** [comment]
@svbergerem: Hmm... [This](https://travis-ci.org/diaspora/diaspora/jobs/197595475#L850) is buzzing me. I don't understand the relation between notifications and comments :/

**Comment by svbergerem** [comment]
@AugierLe42e 
```diff
diff --git a/spec/javascripts/app/views/comment_stream_view_spec.js b/spec/javascripts/app/views/comment_stream_view_spec.js
index cb82a6a7c..b0697b75f 100644
--- a/spec/javascripts/app/views/comment_stream_view_spec.js
+++ b/spec/javascripts/app/views/comment_stream_view_spec.js
@@ -21,10 +21,13 @@ describe("app.views.CommentStream", function(){
     });
 
     it("calls onFormBlur when clicking outside the comment box", function() {
-      spyOn(app.views.CommentStream.prototype, "onFormBlur");
+      // remove existing event handlers
+      $(document.body).off("click");
+
+      spyOn(this.view, "onFormBlur");
       this.view.setupBindings();
       $(document.body).click();
-      expect(app.views.CommentStream.prototype.onFormBlur).toHaveBeenCalled();
+      expect(this.view.onFormBlur).toHaveBeenCalled();
     });
   });
 
diff --git a/spec/javascripts/jasmine_helpers/SpecHelper.js b/spec/javascripts/jasmine_helpers/SpecHelper.js
index 67d167c16..c36a890cf 100644
--- a/spec/javascripts/jasmine_helpers/SpecHelper.js
+++ b/spec/javascripts/jasmine_helpers/SpecHelper.js
@@ -84,6 +84,7 @@ afterEach(function() {
   expect($(".modal-backdrop").length).toBe(0);
   $(".modal-backdrop").remove();
   spec.loadFixtureCount = 0;
+  $(document.body).off();
 });
```

1. Using `spyOn(app.views.CommentStream.prototype, "onFormBlur");` is not wrong, but `spyOn(this.view, "onFormBlur");` works too, because the view already exists.

2. Removing all event handlers from `document.body` after each test fixes your problem. We add an event handler every time we initialize the comment stream view and if we don't remove it then the `onFormBlur` method is called for every fired click event. Even in other tests where `this.commentBox` doesn't exist anymore.

**Comment by AugierLe42e** [comment]
Ah, nice @svbergerem. It's corrected, thanks.

**Comment by Flaburgan** [comment]
Green (apart for this damned block test). Can you confirm it is green locally @AugierLe42e ?

**Comment by AugierLe42e** [comment]
Green locally, @Flaburgan ;)

**Comment by AugierLe42e** [comment]
Ping.

**Comment by svbergerem** [comment]
Pong.

**Comment by AugierLe42e** [comment]
Is it ok to merge?

**Comment by Flaburgan** [comment]
I just did a "user point of view" review and I have no remarks. Nice work! So I'm going to push this on production at diaspora-fr.org to have more user feedback. We'll see if it brings something new.

File: spec/javascripts/app/views/comment_stream_view_spec.js [hunk 15]
```diff
@@ -276,6 +318,72 @@ describe("app.views.CommentStream", function(){
...
+    });
+
+    it("does not call closeForm when the form is clicked", function() {
+      spyOn(this.view, "closeForm");
+      spyOn(this.view.mdEditor, "isPreviewOrTexareaNotEmpty").and.returnValue(false);
+      this.view.onFormBlur($.Event("click", {target: $(".new-comment")}));
+      expect(this.view.closeForm).not.toHaveBeenCalled();
+      this.view.onFormBlur($.Event("click", {target: $(".focus_comment_textarea")}));
+      expect(this.view.closeForm).not.toHaveBeenCalled();
+    });
```
Comment (svbergerem) [inline spec/javascripts/app/views/comment_stream_view_spec.js]: Please add

> it call's closeForm when the user clicks outside of the form

File: app/assets/javascripts/app/views/conversations_form_view.js (hunk 13 above)
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: I'd like to test that this creates a new `MarkdownEditor`.

**Review by svbergerem — CHANGES_REQUESTED** [review]
Most of my remarks about missing tests have been fixed. I reviewed the code again, including the tests. I also checked my other remarks about CSS or behavior bugs. 

- [ ] You fixed the dark theme for comments, but the background for conversations is still wrong. 
- [ ] You also changed the appearance of the empty comment textarea in the dark theme and I'd prefer keeping the old one.
**before**
comment_markdown_editor_before
**after**
comment_markdown_editor_after
- [ ] Autosize for the markdown editor is broken when you open the conversations page and select a conversation afterwards.

**Comment by svbergerem** [comment]
Please note that I also added a comment in the old review: https://github.com/diaspora/diaspora/pull/7214#discussion_r104650950

**Comment by AugierLe42e** [comment]
I don't think I'm going to finish this PR or any other. Feel free to orphan them ;)

**Comment by cmrd-senya** [comment]
@Flaburgan don't you want to take care of finishing this PR?

**Comment by Flaburgan** [comment]
>@Flaburgan don't you want to take care of finishing this PR?

Sure I can do it. Looks like it at least need a rebase at the moment. I just want to be sure this work is the way to go, as I don't know what @svbergerem has in mind / why he didn't answer here at the moment. Maybe he's thinking about another approach.

**Comment by cmrd-senya** [comment]
If you complete this and #7235 I think I'll be able to review them

**Comment by Flaburgan** [comment]
This would close https://github.com/diaspora/diaspora/issues/5210 and https://github.com/diaspora/diaspora/issues/4973

**Comment by denschub** [comment]
Continued in #7482.
